### PR TITLE
EVP: Add EVP_PKEY_get0_provider() and EVP_PKEY_CTX_get0_provider()

### LINKED ISSUE #16058: OpenSSL 3.0 beta 1: How to confirm that an EVP_PKEY operation is being provided by the FIPS module
Section 7.8 of the OpenSSL 3.0 wiki page https://wiki.openssl.org/index.php/OpenSSL_3.0 and https://www.openssl.org/docs/manmaster/man7/fips_module.html each show a similar series of calls that can be used to determine if a message digest operation is being provided by the FIPS module. Equivalent calls also exist for EVP_CIPHER_CTXs. 

However, there does not appear to be an `EVP_PKEY_get0_provider` - type function in OpenSSL 3.0 beta 1 that could be used to follow this chain for an asymmetric operation.  Was this an oversight, or is there a different approach that should be used for asymmetric operations?

### PR DESCRIPTION
**Pull request by levitte**
Fixes #16058

### CODE REVIEW

File: crypto/evp/pmeth_lib.c [hunk 1]
```diff
@@ -1534,6 +1534,28 @@ const char *EVP_PKEY_CTX_get0_propq(EVP_PKEY_CTX *ctx)
...
+    } else if (EVP_PKEY_CTX_IS_KEM_OP(ctx)) {
+        if (ctx->op.encap.kem != NULL)
+            return EVP_KEM_get0_provider(ctx->op.encap.kem);
+    } else if (EVP_PKEY_CTX_IS_ASYM_CIPHER_OP(ctx)) {
+        if (ctx->op.ciph.cipher != NULL)
+            return EVP_ASYM_CIPHER_get0_provider(ctx->op.ciph.cipher);
+    } else if (EVP_PKEY_CTX_IS_GEN_OP(ctx)) {
+        if (ctx->keymgmt != NULL)
+            return EVP_KEYMGMT_get0_provider(ctx->keymgmt);
+    }
```
Comment (paulidale) [inline crypto/evp/pmeth_lib.c:1554]: Should this fall back to the keymgmt if no operation is in progress?

**Review by paulidale — COMMENTED** [review]
Need documentation albeit basic.

File: crypto/evp/pmeth_lib.c (hunk 1 above)
Comment (levitte) [reply crypto/evp/pmeth_lib.c:1554]: I think that would be misleading.

**Comment by levitte** [comment]
> Need documentation albeit basic.

Done

File: doc/man3/EVP_PKEY_is_a.pod [hunk 2]
```diff
@@ -48,6 +57,10 @@ supports signing, otherwise 0.
 
 EVP_PKEY_get0_type_name() returns the name that is found or NULL on error.
 
+EVP_PKEY_get0_description() returns the name that is found or NULL if not.
+
+EVP_PKEY_get0_provider() returns the provider that is found or NULL if not.
```
Comment (paulidale) [inline doc/man3/EVP_PKEY_is_a.pod:62]: s/is/if/

**Review by paulidale — CHANGES_REQUESTED** [review]

File: doc/man3/EVP_PKEY_is_a.pod (hunk 2 above)
Comment (levitte) [reply doc/man3/EVP_PKEY_is_a.pod:62]: Actually, s/that is/if/, no?

Comment (paulidale) [reply doc/man3/EVP_PKEY_is_a.pod]: Yes.  "found" isn't an ideal word either but I can't think of anything better off hand.

File: crypto/evp/pmeth_lib.c [hunk 3]
```diff
@@ -1534,6 +1534,28 @@ const char *EVP_PKEY_CTX_get0_propq(EVP_PKEY_CTX *ctx)
     return ctx->propquery;
 }
 
+const OSSL_PROVIDER *EVP_PKEY_CTX_get0_provider(EVP_PKEY_CTX *ctx)
```
Comment (t8m) [inline crypto/evp/pmeth_lib.c:1537]: Should the parameter be `const EVP_PKEY_CTX *ctx`?

**Comment by t8m** [comment]
Still missing documentation for EVP_PKEY_CTX_get0_provider()

File: crypto/evp/pmeth_lib.c (hunk 3 above)
Comment (levitte) [reply crypto/evp/pmeth_lib.c:1537]: No other function does that...

Comment (paulidale) [reply crypto/evp/pmeth_lib.c:1537]: _We've always done it that way_ isn't a good excuse.

**Comment by levitte** [comment]
> Still missing documentation for EVP_PKEY_CTX_get0_provider()

Added

**Comment by t8m** [comment]
CI failure is relevant

File: crypto/evp/pmeth_lib.c (hunk 3 above)
Comment (t8m) [reply crypto/evp/pmeth_lib.c:1537]: > No other function does that...

Yeah, I've noticed. Is it too late for constification of the other getter functions for 3.0? Adding const in such cases does not break API and could be regarded as a bug fix in some sense.

**Comment by levitte** [comment]
> CI failure is relevant

One little comma!  Argh...

... at least, that one didn't blow up a rocket :wink:

**Review by t8m — APPROVED** [review]

File: crypto/evp/pmeth_lib.c (hunk 3 above)
Comment (paulidale) [reply crypto/evp/pmeth_lib.c:1537]: I don't think it's too late but it would probably require vote...
I'd still like to see a beginning here.

**Comment by paulidale** [comment]
Is the constification going to be added or not?

**Comment by levitte** [comment]
> Is the constification going to be added or not?

I'll include the commit that did so

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
Merged

ad0a2c011020268a242737820bc50549e76cd6b8 EVP: Add EVP_PKEY_get0_provider() and EVP_PKEY_CTX_get0_provider()
317ed1b41790db7187bc6585b3d57c6a983c793a DOCS: Move the description of EVP_PKEY_get0_description()

**Pull request closed without merging**